kernels/rviv/3/bli_sdgemm_rviv_asm_4vx4.inc

```diff
@@ -0,0 +1,396 @@
+#define BETA   fa2
+
+#define B00    fa4
+#define B01    fa5
+#define B02    fa6
+#define B03    fa7
+
+#define B10    fa0
+#define B11    fa1
+#define B12    fa2
+#define B13    fa3
+
+#define A00    v24
+#define A10    v25
+#define A20    v26
+#define A30    v27
+
+#define A01    v28
+#define A11    v29
+#define A21    v30
+#define A31    v31
+
+#define AB00   v0
+#define AB01   v1
+#define AB02   v2
+#define AB03   v3
+#define AB10   v4
+#define AB11   v5
+#define AB12   v6
+#define AB13   v7
+#define AB20   v8
+#define AB21   v9
+#define AB22   v10
+#define AB23   v11
+#define AB30   v12
+#define AB31   v13
+#define AB32   v14
+#define AB33   v15
+
+#define rs_c   a6
+#define cs_c   a7
+
+REALNAME:
+    #include "rviv_save_registers.inc"
+
+    // Arbitrarily long length
+    // TODO: Find a better way of querying MAXVL
+    li tmp, 256
+    vsetvli s0, tmp, VTYPE, m1, ta, ma
+    slli s0, s0, DATASHIFT      // vector length in bytes
+
+    mv rs_c, s0
+    slli cs_c, cs_c, DATASHIFT  // count in FP steps
+
+    // Set up pointers
+    add C01_ptr, C00_ptr, cs_c
+    add C02_ptr, C01_ptr, cs_c
+    add C03_ptr, C02_ptr, cs_c
+    add C10_ptr, C00_ptr, rs_c
+    add C11_ptr, C01_ptr, rs_c
+    add C12_ptr, C02_ptr, rs_c
+    add C13_ptr, C03_ptr, rs_c
+
+    add A1_ptr, A0_ptr, s0
+    add A2_ptr, A1_ptr, s0
+    add A3_ptr, A2_ptr, s0
+
+    // Zero-initialize accumulators
+    vxor.vv AB00, AB00, AB00
+    vxor.vv AB01, AB01, AB01
+    vxor.vv AB02, AB02, AB02
+    vxor.vv AB03, AB03, AB03
+    vxor.vv AB10, AB10, AB10
+    vxor.vv AB11, AB11, AB11
+    vxor.vv AB12, AB12, AB12
+    vxor.vv AB13, AB13, AB13
+    vxor.vv AB20, AB20, AB20
+    vxor.vv AB21, AB21, AB21
+    vxor.vv AB22, AB22, AB22
+    vxor.vv AB23, AB23, AB23
+    vxor.vv AB30, AB30, AB30
+    vxor.vv AB31, AB31, AB31
+    vxor.vv AB32, AB32, AB32
+    vxor.vv AB33, AB33, AB33
+
+    slli s0, s0, 2 // length of a column of A in bytes
+
+    li tmp, 1
+    ble loop_counter, tmp, TAIL
+
+    // Load B(l,0:3)
+    FLOAD B00, 0*DATASIZE(B_row_ptr)
+    FLOAD B01, 1*DATASIZE(B_row_ptr)
+    FLOAD B02, 2*DATASIZE(B_row_ptr)
+    FLOAD B03, 3*DATASIZE(B_row_ptr)
+
+    // Load A(:,l)
+    VLE A00, (A0_ptr)
+    VLE A10, (A1_ptr)
+    VLE A20, (A2_ptr)
+    VLE A30, (A3_ptr)
+
+LOOP_K:
+    addi loop_counter, loop_counter, -2
+
+    // Point to A(:,l+1)
+    add A0_ptr, A0_ptr, s0
+    add A1_ptr, A1_ptr, s0
+    add A2_ptr, A2_ptr, s0
+    add A3_ptr, A3_ptr, s0
+
+    vfmacc.vf AB00, B00, A00
+    vfmacc.vf AB01, B01, A00
+    vfmacc.vf AB02, B02, A00
+    vfmacc.vf AB03, B03, A00
+
+    vfmacc.vf AB10, B00, A10
+    vfmacc.vf AB11, B01, A10
+    vfmacc.vf AB12, B02, A10
+    vfmacc.vf AB13, B03, A10
+
+    // Load B(l+1,0:3)
+    FLOAD B10, 4*DATASIZE(B_row_ptr)
+    FLOAD B11, 5*DATASIZE(B_row_ptr)
+    FLOAD B12, 6*DATASIZE(B_row_ptr)
+    FLOAD B13, 7*DATASIZE(B_row_ptr)
+    addi B_row_ptr, B_row_ptr, 8*DATASIZE
+
+    // Load A(:,l+1)
+    VLE A01, (A0_ptr)
```

Comment:
nit: Not need to have a macro here, `vle.v A01, (A0_ptr)` picks up the right version of vle (vle32.v, vle64.v)

Reply:
```
kernels/rviv/3/bli_sdgemm_rviv_asm_4vx4.inc:161: Error: unrecognized opcode `vle.v v30,(t1)'

kernels/rviv/3/bli_sdgemm_rviv_asm_4vx4.inc:302: Error: unrecognized opcode `vse.v v9,(t3)'
```

Reply:
Interesting...another compiler problem

Reply:
I suggest keeping the macros for now.